TypeScript: Fix incorrectly removes double parentheses around types

sosukesuzuki opened the pull request:
<!-- Please provide a brief summary of your changes: -->

Fix #6603 

If we don't want to add this PR to the 1.19 milestone, we can add it to the 1.20 milestone without merging now.

```ts
// Input
type A = 0 extends ((1 extends 2  ? 3 : 4)) ? 5 : 6;
type B = ((0 extends 1 ? 2 : 3)) extends 4 ? 5 : 6:
type C = ((number | string))["toString"];
type D = ((keyof T1))["foo"];

// Prettier (stable)
type A = 0 extends 1 extends 2 ? 3 : 4 ? 5 : 6;
type B = 0 extends 1 ? 2 : 3 extends 4 ? 5 : 6;
type C = number | string["toString"];
type D = keyof T1["foo"];

// Prettier (master)
type A = 0 extends (1 extends 2 ? 3 : 4) ? 5 : 6;
type B = (0 extends 1 ? 2 : 3) extends 4 ? 5 : 6;
type C = (number | string)["toString"];
type D = (keyof T1)["foo"];
```

<!-- Please ensure you’ve done all of these things (if applicable). -->
<!-- You can replace the `[ ]` with `[x]` to mark each task as done. -->

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

uhyo commented:
Thank you for your quick action! I'm the reporter of the problem.

But I found another case to consider: [playground link](https://deploy-preview-6604--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAKgRmwF5sAKUqAVwFsAjOAJ2wB9sBnGBgSygHMBKftgBk2WhAgAbOAEMoAbhAAaEBAwwu0NslAyGDCAHcACnoTaUMgG4QuAE2UhaDGWADWcGAGUMrnr2ROSjgVAAsYakkAdVCueDZfMDgvcziuKzi0ZHA2bRUeNkYYYxdeahlkADMZSUKVACs2AA8AIRd3Ty8ZajgAGR44Kpq6kEamr39pAEVKCHgh2pCQXwZChmz0LDYwbnVHDG5YKPsYUOQADgAGFQOIQqiXDGyDuDWrQZUAR1n4ErULEAyNgAWigcDgdghjgYcG+XBhJRkZQqSGqixUhWoXECDGCGMmcBmc0GqOGSxgMloxzsp2QACYVJwZFxJP4AMIQajlbJQaAfECUQq4SkWNEjKzBACSUEhsC8Oy46gAgjKvOhpAtCgBfLVAA)

```ts
// Input
type T1 = ((number | string)) & boolean;

// Output
type T1 = number | string & boolean;
```

evilebottnawi commented:
@sosukesuzuki good job!

uhyo commented:
It seems that we have to handle unions too: [playground link](https://deploy-preview-6604--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEMCeAHOACA8ga2wF5soBXAWwCM4AnbAH2wApmBDWgcyWzajQCUxAHzYAbhACWAEwEAdWJhwAhMp2KlKNek1bsuPPoJHipsgSAA0ICBhiToAZ2SgOtCAHcAChwTOUbBIyViBUtGxg+HAwAMoYEZJQnMgwtGRw1gAWMBQANgDqmZLwjvFgcDF+xZJixWjI4I7O1omOdDBe4ZwUbMgAZmy5bdYAVo4AHsrhkdExbBRwADKJcP2DwyBj4zGJnLlwAIpkEPBrQxkg8bRttA3oWI5gtJJ2IRjPsPkyMJnIABwABms7wgbXy4QwDXecBuYlW1gAjsd4J1bP4QGxHABaKBwODSfEhWhwJGSYmdNjdXpIAbnaxtCiSFJpC6OXb7I4nVY09YXGBsKhfaQ-ZAAJmsqTYklyuwAwhAKD0GlBoPCQGQ2gAVAX+WkbMTpACSUAJsBiTxeMAAgiaYuh9mc2gBfJ1AA)

```ts
type Ok = number | ((arg: any) => void)
type Bug = number | (((arg: any) => void))
```

evilebottnawi commented:
/cc @thorn0

thorn0 commented:
@sosukesuzuki Would you mind if I use the tests you wrote here for #6605?

thorn0 commented:
It's probably better not to change the tests in `tests/typescript/conformance`. They were taken [from the TS repo](https://github.com/microsoft/TypeScript/tree/master/tests/cases/conformance), and potentially we are interested in keeping them more or less in sync as TS evolves.

evilebottnawi commented:
/cc @sosukesuzuki can we rebase and fix problem above? thanks

sosukesuzuki commented:
@thorn0 @evilebottnawi 
I rebased and fixed problem above. But, the directory for testing intersection types does not exist so I added new directory `typescript-intersection`. Please let me know if there is a problem.

thorn0 commented:
Looks good.
